Compiling SuperSQL query causes "panic: runtime error: index out of range [-1]"

## tl;dr

```
$ super -version
Version: v1.18.0-304-g6300fbaf

$ super -z -c "
SELECT dt,
       hr,
       SUM(net_in) AS net_in_sum,
       SUM(net_out) AS net_out_sum,
       SUM(net_in) + SUM(net_out) AS both_sum
FROM (
  SELECT bucket(log_time, 1h) AS hr,
         COALESCE(bytes_in, 0.0) / 1000000000.0 AS net_in,
         COALESCE(bytes_out, 0.0) / 1000000000.0 AS net_out
  FROM 'foobar'
  WHERE machine_name IN ('allsorts','andes','bigred','blackjack','bonbon',
      'cadbury','chiclets','cotton','crows','dove','fireball','hearts','huey',
      'lindt','milkduds','milkyway','mnm','necco','nerds','orbit','peeps',
      'poprocks','razzles','runts','smarties','smuggler','spree','stride',
      'tootsie','trident','wrigley','york')
) AS r
GROUP BY dt,
         hr
ORDER BY both_sum DESC
LIMIT 10"
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/brimdata/super/pkg/field.Path.Leaf(...)
	/Users/phil/work/super/pkg/field/field.go:27
github.com/brimdata/super/compiler/semantic.inferColumnName({0x12a1a460?, 0xc000210750?}, {0x12a2d3a8, 0xc0003b1c50})
	/Users/phil/work/super/compiler/semantic/sql.go:611 +0x6c
github.com/brimdata/super/compiler/semantic.(*analyzer).semAs(0xc0005cd870, {0x12a351d8?, 0xc0001168a0?}, {{0x12304a20, 0x6}, 0x0, {0x12a2d3a8, 0xc0003b1c50}, {0x8, 0x9}}, ...)
	/Users/phil/work/super/compiler/semantic/sql.go:584 +0x7d
github.com/brimdata/super/compiler/semantic.(*analyzer).semProjection(0xc0005cd870, 0xc0001168a0, {0xc000660000, 0x5, 0x0?}, 0xc0005ccaf8)
	/Users/phil/work/super/compiler/semantic/sql.go:561 +0x305
github.com/brimdata/super/compiler/semantic.(*analyzer).semSelect(0xc0005cd870, 0xc00043b040, {0x0?, 0xc0001ae840?, 0xc0000a58c0?})
	/Users/phil/work/super/compiler/semantic/sql.go:39 +0x145
github.com/brimdata/super/compiler/semantic.(*analyzer).semSQLOp(0xc0005cd870, {0x12a2e788?, 0xc00043b040}, {0x0, 0xc0005ccd08?, 0x0})
	/Users/phil/work/super/compiler/semantic/sql.go:367 +0xfa
github.com/brimdata/super/compiler/semantic.(*analyzer).semSQLOp(0xc0005cd870, {0x12a2e6c8?, 0xc0006cb5e0}, {0x0, 0x2?, 0x0})
	/Users/phil/work/super/compiler/semantic/sql.go:376 +0x82b
github.com/brimdata/super/compiler/semantic.(*analyzer).semSQLOp(0xc0005cd870, {0x12a2e6f8?, 0xc0006cbcc0}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/sql.go:407 +0x4cc
github.com/brimdata/super/compiler/semantic.(*analyzer).semOp(0xc0005cd870, {0x12a2e6f8?, 0xc0006cbcc0}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/op.go:555 +0x2992
github.com/brimdata/super/compiler/semantic.(*analyzer).semSeq(0xc0005cd870, {0xc000169a20, 0x1, 0x1})
	/Users/phil/work/super/compiler/semantic/op.go:33 +0x51
github.com/brimdata/super/compiler/semantic.Analyze({0x12a34b40, 0xc0001f6690}, 0xc000025f00, 0xc000116020, 0x0)
	/Users/phil/work/super/compiler/semantic/analyzer.go:21 +0x215
github.com/brimdata/super/compiler.Analyze(...)
	/Users/phil/work/super/compiler/package.go:30
github.com/brimdata/super/compiler.CompileWithAST(0xc0001f6690, 0x10c566c5?, 0xc000116020, 0x1, 0xc, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/package.go:77 +0x65
github.com/brimdata/super/compiler.(*compiler).NewQuery(0x12a34be8?, 0xc00042e870?, 0xc0005cd9d0?, {0x0?, 0x8?, 0x127eaf40?}, 0x10b80720?)
	/Users/phil/work/super/compiler/compiler.go:38 +0x45
github.com/brimdata/super/runtime.CompileQuery({0x12a34be8?, 0xc00042e870?}, 0xc0006cbd60, {0x12a22df0, 0xc00042a148}, 0xc000025f00, {0x0, 0x0, 0x0})
	/Users/phil/work/super/runtime/compiler.go:34 +0x10f
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc000470200, {0xc00003c0f0, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:158 +0x4c9
github.com/brimdata/super/pkg/charm.path.run({0xc00042a2a0, 0x1, 0x1}, {0xc00003c0f0, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7b
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0x13b52a60, {0xc00003c0d0, 0x3, 0x3})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:40 +0x5b
```

## Details

Repro is with super commit 6300fba.

The query above was an iteration in my mgbench work where I was trying to rewrite the [bench1/q6 SPQ](https://github.com/brimdata/mgbench-private/blob/main/bench1/q6/q6.spq) to be closer to the original [bench1/q6 SQL](https://github.com/brimdata/mgbench-private/blob/main/bench1/q6/q6.sql) by taking advantage of the many enhancements in #5669.

Note: I've observed that no input data is needed to repro this problem. While my scripts were originally trying to query the genuine test data, I stumbled onto the fact that I could use a garbage `FROM` such as the `'foobar'` shown above and the panic is exactly the same.

--- a/compiler/semantic/schema.go
+++ b/compiler/semantic/schema.go
@@ -131,7 +131,7 @@ func (s *staticSchema) resolveColumn(col string) (field.Path, error) {
 			return field.Path{col}, nil
 		}
 	}
-	return nil, fmt.Errorf("%q: column does not exist", col)
+	return nil, fmt.Errorf("column %q: does not exist", col)
 }
 
 func (a *anonSchema) resolveColumn(col string) (field.Path, error) {
@@ -140,52 +140,38 @@ func (a *anonSchema) resolveColumn(col string) (field.Path, error) {
 			return field.Path{col}, nil
 		}
 	}
-	return nil, nil
+	return nil, fmt.Errorf("column %q: does not exist", col)
 }
 
 func (s *selectSchema) resolveColumn(col string) (field.Path, error) {
 	if s.out != nil {
-		resolved, err := s.out.resolveColumn(col)
-		if err != nil {
-			return nil, err
-		}
-		if resolved != nil {
+		if resolved, _ := s.out.resolveColumn(col); resolved != nil {
 			return append([]string{"out"}, resolved...), nil
 		}
 	}
 	resolved, err := s.in.resolveColumn(col)
-	if err != nil {
-		return nil, err
-	}
 	if resolved != nil {
 		return append([]string{"in"}, resolved...), nil
 	}
-	return nil, nil
+	return nil, err
 }
 
 func (j *joinSchema) resolveColumn(col string) (field.Path, error) {
-	out, err := j.left.resolveColumn(col)
-	if err != nil {
-		return nil, err
-	}
-	if out != nil {
-		chk, err := j.right.resolveColumn(col)
-		if err != nil {
-			return nil, err
-		}
-		if chk != nil {
+	left, lerr := j.left.resolveColumn(col)
+	if left != nil {
+		if chk, _ := j.right.resolveColumn(col); chk != nil {
 			return nil, fmt.Errorf("%q: ambiguous column reference", col)
 		}
-		return append([]string{"left"}, out...), nil
+		return append([]string{"left"}, left...), nil
 	}
-	out, err = j.right.resolveColumn(col)
-	if err != nil {
-		return nil, err
+	if lerr == nil {
+		panic("TBD") //XXX new contract is that nil result requires err
 	}
-	if out != nil {
-		return append([]string{"right"}, out...), nil
+	right, rerr := j.right.resolveColumn(col)
+	if right != nil {
+		return append([]string{"right"}, right...), nil
 	}
-	return nil, nil
+	return nil, fmt.Errorf("%q: not found (%w, %w)", col, lerr, rerr)
 }
 
 func (d *dynamicSchema) deref(name string) (dag.Expr, schema) {

--- a/compiler/semantic/schema.go
+++ b/compiler/semantic/schema.go
@@ -166,7 +166,7 @@ func (j *joinSchema) resolveColumn(col string) (field.Path, error) {
 	}
 	if lerr == nil {
 		// This shouldn't happen because the resolve return values should
-		// always be nil/err or val/err.
+		// always be nil/err or val/nil.
 		panic("issue encountered in SQL name resolution")
 	}
 	right, rerr := j.right.resolveColumn(col)

--- a/compiler/semantic/schema.go
+++ b/compiler/semantic/schema.go
@@ -131,7 +131,7 @@ func (s *staticSchema) resolveColumn(col string) (field.Path, error) {
 			return field.Path{col}, nil
 		}
 	}
-	return nil, nil
+	return nil, fmt.Errorf("%q: column does not exist", col)
 }
 
 func (a *anonSchema) resolveColumn(col string) (field.Path, error) {

--- a/compiler/semantic/schema.go
+++ b/compiler/semantic/schema.go
@@ -165,7 +165,9 @@ func (j *joinSchema) resolveColumn(col string) (field.Path, error) {
 		return append([]string{"left"}, left...), nil
 	}
 	if lerr == nil {
-		panic("TBD") //XXX new contract is that nil result requires err
+		// This shouldn't happen because the resolve return values should
+		// always be nil/err or val/err.
+		panic("issue encountered in SQL name resolution")
 	}
 	right, rerr := j.right.resolveColumn(col)
 	if right != nil {

--- a/compiler/semantic/scope.go
+++ b/compiler/semantic/scope.go
@@ -142,7 +142,7 @@ func resolvePath(sch schema, path field.Path) (field.Path, error) {
 	if table != nil {
 		columnPath, err := table.resolveColumn(path[1])
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("table %q: %w", path[0], err)
 		}
 		if columnPath != nil {
 			out := append(tablePath, columnPath...)

--- a/compiler/semantic/sql.go
+++ b/compiler/semantic/sql.go
@@ -426,7 +426,10 @@ func (a *analyzer) semSQLJoin(join *ast.SQLJoin, seq dag.Seq) (dag.Seq, schema) 
 	sch := &joinSchema{left: leftSchema, right: rightSchema}
 	leftKey, rightKey, err := a.semSQLJoinCond(join.Cond, sch)
 	if err != nil {
-		a.error(join.Cond, errors.New("SQL joins currently limited to equijoin on fields"))
+		// Join expression errors are already logged so suppress further notice.
+		if err != badJoinCond {
+			a.error(join.Cond, err)
+		}
 		return append(seq, badOp()), badSchema()
 	}
 	assignment := dag.Assignment{
@@ -450,6 +453,8 @@ func (a *analyzer) semSQLJoin(join *ast.SQLJoin, seq dag.Seq) (dag.Seq, schema) 
 	return append(append(leftSeq, par), dagJoin), sch
 }
 
+var badJoinCond = errors.New("bad join condition")
+
 func (a *analyzer) semSQLJoinCond(cond ast.JoinExpr, sch *joinSchema) (*dag.This, *dag.This, error) {
 	//XXX we currently require field expressions for SQL joins and will need them
 	// to resolve names to join side when we add scope tracking
@@ -462,15 +467,29 @@ func (a *analyzer) semSQLJoinCond(cond ast.JoinExpr, sch *joinSchema) (*dag.This
 	if err != nil {
 		return nil, nil, err
 	}
-	left, ok := l.(*dag.This)
-	if !ok {
-		return nil, nil, errors.New("join keys must be field references")
+	left, err := joinFieldAsThis("left", l)
+	if err != nil {
+		return nil, nil, err
 	}
-	right, ok := r.(*dag.This)
-	if !ok {
-		return nil, nil, errors.New("join keys must be field references")
+	right, err := joinFieldAsThis("right", r)
+	return left, right, err
+}
+
+func joinFieldAsThis(which string, e dag.Expr) (*dag.This, error) {
+	if _, ok := e.(*dag.BadExpr); ok {
+		return nil, badJoinCond
 	}
-	return left, right, nil
+	this, ok := e.(*dag.This)
+	if !ok {
+		return nil, errors.New("join condition must be equijoin on fields")
+	}
+	if len(this.Path) == 0 {
+		return nil, errors.New("join expression cannot refer to 'this'")
+	}
+	if this.Path[0] != which {
+		return nil, fmt.Errorf("%s-hand side of join condition must refer to %s-hand table", which, which)
+	}
+	return this, nil
 }
 
 func (a *analyzer) semJoinCond(cond ast.JoinExpr) (dag.Expr, dag.Expr, error) {

--- a/compiler/ztests/sql/join-static.yaml
+++ b/compiler/ztests/sql/join-static.yaml
@@ -1,0 +1,13 @@
+script: |
+  super -z -c "select * from (select 1 as a) join (select 1 as b) on a=b"
+  super -z -c "select * from (select 1 as a) t1 join (select 1 as b) on t1.a=b"
+  super -z -c "select * from (select 1 as a) join (select 1 as b) t2 on a=t2.b"
+  super -z -c "select * from (select 1 as a) t1 join (select 1 as a) t2 on t1.a=t2.a"
+
+outputs:
+  - name: stdout
+    data: |
+      {a:1,b:1}
+      {a:1,b:1}
+      {a:1,b:1}
+      {a:1}

--- a/compiler/ztests/sql/scope-errors.yaml
+++ b/compiler/ztests/sql/scope-errors.yaml
@@ -1,0 +1,33 @@
+script: |
+  ! super -z -c "select * from (select 1 as a, 2 as b) t1 join (select 2 as b) on a=b"
+  echo === 1>&2
+  ! super -z -c "select * from a.json join b.json on a=b"
+  echo === 1>&2
+  ! super -z -c "select * from (select 2 as x) as a join b.json on x=b.c"
+  echo === 1>&2
+  ! super -z -c "select * from (select 2 as x) a join b.json a on a.x=a.y"
+
+outputs:
+  - name: stderr
+    data: |
+      "b": ambiguous column reference at line 1, column 68:
+      select * from (select 1 as a, 2 as b) t1 join (select 2 as b) on a=b
+                                                                         ~
+      ===
+      "a": ambiguous column reference at line 1, column 37:
+      select * from a.json join b.json on a=b
+                                          ~
+      "b": ambiguous column reference at line 1, column 39:
+      select * from a.json join b.json on a=b
+                                            ~
+      ===
+      "x": ambiguous column reference at line 1, column 51:
+      select * from (select 2 as x) as a join b.json on x=b.c
+                                                        ~
+      ===
+      "a": ambiguous table reference at line 1, column 50:
+      select * from (select 2 as x) a join b.json a on a.x=a.y
+                                                       ~~~
+      "a": ambiguous table reference at line 1, column 54:
+      select * from (select 2 as x) a join b.json a on a.x=a.y
+                                                           ~~~

--- a/compiler/ztests/sql/select-nonexistent-column.yaml
+++ b/compiler/ztests/sql/select-nonexistent-column.yaml
@@ -1,0 +1,7 @@
+zed: |
+  SELECT x FROM (SELECT 1)
+
+error: |
+  "x": column does not exist at line 1, column 8:
+  SELECT x FROM (SELECT 1)
+         ~

--- a/compiler/ztests/sql/select-nonexistent-column.yaml
+++ b/compiler/ztests/sql/select-nonexistent-column.yaml
@@ -2,6 +2,6 @@ zed: |
   SELECT x FROM (SELECT 1)
 
 error: |
-  "x": column does not exist at line 1, column 8:
+  column "x": does not exist at line 1, column 8:
   SELECT x FROM (SELECT 1)
          ~

--- a/compiler/ztests/sql/select-nonexistent-table-column.yaml
+++ b/compiler/ztests/sql/select-nonexistent-table-column.yaml
@@ -1,0 +1,7 @@
+zed: |
+  SELECT t.x FROM (SELECT 1 as y) t
+
+error: |
+  table "t": column "x": does not exist at line 1, column 8:
+  SELECT t.x FROM (SELECT 1 as y) t
+         ~~~

